##Phase by Rare Variants
This program performs phasing of Whole Genome Sequencing genotypes in pedigrees utilizing lineage specific alleles to establish the IBD pattern within the pedigrees.

###Convert_012_to_HIPster_format.pl
* This program converts 012 format output by VCFtools to HIPster format
* Usage is as follows: >perl Convert_012_to_HIPster_format.pl $ARGV[0]
* where $ARGV[0] is the base name of the output files generated by VCFtools when converting to 012 format 

###Convert_HIPster_to_PED_MAP.pl
* This program converts HIPster format to PED/MAP format for use by SHAPEIT
* Usage is as follows: >perl Convert_HIPster_to_PED_MAP.pl $ARGV[0] $ARGV[1] $ARGV[2] $ARGV[3]
* where $ARGV[0] is the file of genotypes in HIPster format, $ARGV[1] is a name for the map file, $ARGV[2] is the name of the pedigree file in HIPster format, and $ARGV[3] is a name for the ped file

###Convert_IMPUTE_to_HIPster.pl
* This program converts *.haps file from SHAPEIT to HIPSTER format
* Usage is as follows: >perl Convert_IMPUTE_to_HIPster.pl $ARGV[0] $ARGV[1]
* where $ARGV[0] is the .ped file for the pedigree phased by SHAPEIT and $ARGV[1] is .haps file

###Merge_results.pl
* This program merges phasing results from PBRV and SHAPEIT
* Usage is as follows: >perl Merge_results.pl $ARGV[0] $ARGV[1]
* where $ARGV[0] is the Phased_chromosomes.csv file produced by PBRV and $ARGV[1] .haps file from SHAPEIT that has been converted to HIPster format

###PBRV.pl
* This program performs phasing of Whole Genome Sequencing genotypes in pedigrees utilizing lineage specific alleles to establish the IBD pattern within the pedigrees.
* Usage is as follows: >perl Merge_results.pl $ARGV[0] $ARGV[1] $ARGV[1]
* where $ARGV[0] is the pedigree file in HIPster format, $ARGV[1] is the genotype file in HIPster format, and $ARGV[2] is an integer above 0 that is used as the number of adjacent observations that are required to identify a recombination event (we suggest 5)

###Calculate_metrics.pl
* This program calculates switch error rate among other metrics given the phasing results and known haplotypes.
* Usage is as follows: >perl Calculate_metrics.pl $ARGV[0] $ARGV[1] $ARGV[1]
* Where $ARGV[0] is the phasing results, $ARGV[1] is the known haplotypes, and $ARGV[2] is the set of diploid genotypes given to the algorithm

###Example/
* This folder provides an example for a small pedigree. All_Diploid_genos
* To run the example: >perl PBRV.pl Seven_sib_pedigree.csv Diploid_genos.csv 5
